java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageReader.java

```diff
@@ -18,48 +18,50 @@
 
 package org.apache.arrow.vector.ipc.message;
 
-
-import io.netty.buffer.ArrowBuf;
-import org.apache.arrow.flatbuf.Message;
-import org.apache.arrow.memory.BufferAllocator;
-
 import java.io.IOException;
 
 /**
- * Interface for reading a sequence of messages.
+ * Abstract base class for reading a sequence of messages.
  */
-public interface MessageReader {
+public abstract class MessageReader<T extends MessageReadHolder> {
 
   /**
-   * Read the next message in the sequence.
+   * Read the next message in the sequence and message body, if message has a defined
+   * body length.
    *
-   * @return The read message or null if reached the end of the message sequence
+   * @param holder Data structure to hold message information read.
+   * @return true if a Message was read or false if no more messages.
    * @throws IOException
    */
-  Message readNextMessage() throws IOException;
+  public boolean readNext(T holder) throws IOException {
 
-  /**
-   * When a message is followed by a body of data, read that data into an ArrowBuf. This should
-   * only be called when a Message has a body length > 0.
-   *
-   * @param message Read message that is followed by a body of data
-   * @param allocator BufferAllocator to allocate memory for body data
-   * @return An ArrowBuf containing the body of the message that was read
-   * @throws IOException
-   */
-  ArrowBuf readMessageBody(Message message, BufferAllocator allocator) throws IOException;
+    // Read message and body if bodyLength is specified
+    readMessage(holder);
+    if (holder.message == null) {
+      return false;
+    }
+    if (holder.message.bodyLength() > 0) {
+      readMessageBody(holder);
+    }
+    return true;
+  }
 
   /**
-   * Return the current number of bytes that have been read.
+   * Attempt to read a message. If no new message is available, set holder.message to null to
+   * indicate no further messages in the sequence.
    *
-   * @return number of bytes read
+   * @param holder Message and message information that is populated when read by implementation.
+   * @throws IOException
    */
-  long bytesRead();
+  protected abstract void readMessage(T holder) throws IOException;
```

Comment:
I don't think we should separate readMessage and readMessageBody as separate methods. The holder can hold both. Let's just have a single boolean readMessage(T holder) that returns false if there are no more messages.

Reply:
I'm ok with that, but I believe @icexelloss had an earlier concern that when implementing a custom MessageReader, the subclass would then have to be aware of some details of the Arrow stream format. Specifically, checking for the EOS and if the message is followed by a body or not.

Also, I'm not sure how much value having an interface with just `boolean readMessage(T holder)` since the implementation will have to do everything and we could not use this interface for `ArrowStreamReader` since that requires a `ReadChannel` and returning the message body in an `ArrowBuf`.

Reply:
I think what is useful for someone writing a custom `MessageReader` would be to have a hook after reading the message to inspect the message type and then some way to override the reading of the body in case they do not want it to be read into an `ArrowBuf`.  In my case, I need to check for a `RecordBatch` message and then forward the body data somewhere else. I'll think about this a little more..